waku/v2/protocol/waku_store/waku_store.nim

```diff
@@ -612,14 +612,23 @@ proc queryFromWithPaging*(w: WakuStore, query: HistoryQuery, peer: RemotePeerInf
   return ok(messageList)
 
 proc queryLoop(w: WakuStore, query: HistoryQuery, candidateList: seq[RemotePeerInfo]): Future[MessagesResult]  {.async, gcsafe.} = 
-  ## loops through the candidateList in order and sends the query to each until one of the query gets resolved successfully
-  ## returns the retrieved messages, or error if all the requests fail
-  for peer in candidateList.items: 
-    let successResult = await w.queryFromWithPaging(query, peer)
-    if successResult.isOk: return ok(successResult.value)
-
-  debug "failed to resolve the query"
-  return err("failed to resolve the query")
+  ## loops through the candidateList in order and sends the query to each
+  ## once all responses have been received, the retrieved messages are consolidated into one deduplicated list
+  ## if no messages have been retrieved, the returned future will resolve into a MessagesResult result holding an empty seq.
+  var futureList: seq[Future[MessagesResult]]
+  for peer in candidateList.items:
+    futureList.add(w.queryFromWithPaging(query, peer))
+  futureList = await allFinished(futureList) # TODO: is there a proc that returns a Future[seq[T]?
```

Comment:
Any advantage to doing this rather than just use `await allFutures()`? Seems like that will at least avoid double-assigning to `futureList`? It will still be necessary to retrieve the completed values iteratively, though, as you did below.

Reply:
Thank you. I will apply this change. As I am very new to Nim, I did not know better 😅. At first I thought I could save the following iteration by directly getting a list with the results of all successfully resolved futures, and then replace the allFInished call with the respective call. But as you pointed out, using allFinished, I still have to iterate.